histogram: waveform processing/drawing fixes/cleanups

dtorop opened the pull request:
This is a general cleanup of the waveform processing/drawing code. The aim is to:

1. Properly fix #7307, #7429
2. Use some of the nice new image handling code/techniques for @ralfbrown to clarify/cleanup the code.

The waveform processing loop now iterates explicitly by bin. It doesn't rely on OpenMP directives to enforce bin-by-bin processing. As noted in https://github.com/darktable-org/darktable/issues/7307#issuecomment-757090720, that will fail in the case of a `schedule(simd:static ...)`.

Also are using simpler OpenMP clauses or skipping OpenMP entirely when multiple threads for a small buffer don't show a speedup.

This code should be approximately the same speed as the code currently in master.

@MStraeten : I haven't tested this on OS X. If you happened to have time to compile/test this, I'd be quite interested to know how it compares to master after your #7518. Ideally, it should be faster and not regress back to the "strange" waveform histogram.

dtorop commented:
I did slightly better benchmarks. The waveform processing is about 30% slower with this PR. The waveform drawing code is about 30% faster. Net, at least on this machine, the code is faster, as for each process step (0.002 sec. slowdown), there's at least one draw step (0.006 sec. speedup). The new code also no longer relies on CPU-dependent OpenMP behavior.

step | time (wall) | time (CPU)
---- | ------------- | --------------
waveform process (master) | 0.007 | 0.053
waveform process (this PR) | 0.009 | 0.066
waveform draw (master) | 0.019 | 0.024
waveform draw (this PR) | 0.013 | 0.017

MStraeten commented:
built on osx - histogram is fine.
performance:
quick&dirty fix current master: `final histogram waveform took 0.001 secs (0.009 CPU)`
this pr: `final histogram waveform took 0.004 secs (0.017 CPU)`

dtorop commented:
@MStraeten: Thank you for testing so quickly.

Two more questions, if you had time:

1. What are the "scope" draw timings before & after this PR? This code touched those as well. They should be something like: `scope draw took 0.012 secs (0.010 CPU)` when in waveform mode.
2. What are the waveform process timings with the new code with OpenMP disabled?

To test the latter, the same change as your #7518:

```diff
modified   src/libs/histogram.c
@@ -218,8 +218,8 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
 
   // count the colors
   // FIXME: could flip x/y axes here and when reading to make row-wise iteration?
-#if defined(_OPENMP)
-#pragma omp parallel for default(none) \
+#if defined(_OPENMP) && !defined(__APPLE__)
+#pragma omp parallel for default(none)                                  \
   dt_omp_firstprivate(input, wf_linear, width, height, wf_width, bin_width, _height, scale) \
   schedule(static)
 #endif
```

What I get from your test so far:

- This looks not to be a compiler bug, but MacOS clang doing something unexpected -- as allowed by OpenMP spec -- with a badly specified directive.
- It's possible this goes for #7349 as well.

MStraeten commented:
q&d: 
initial open of the image in darkroom: `scope draw took 0.122 secs (0.257 CPU)`
then for each compress history: `scope draw took 0.014 secs (0.121 CPU)`
pr:
initial open: `scope draw took 0.166 secs (1.946 CPU)`
each compress history: `scope draw took 0.014 secs (0.107 CPU)`
pr without openmp:
initial open: `scope draw took 0.119 secs (0.261 CPU)`
each compress history: `scope draw took 0.014 secs (0.099 CPU)`

dtorop commented:
Thank you! What is the "final histogram waveform" time without OpenMP in `_lib_histogram_process_waveform()`? And are the last two times you sent with OpenMP disabled in `_lib_histogram_process_waveform()` or in `_lib_histogram_draw_waveform_channel()`? The former is the function which would change "final histogram waveform" time, the latter changes "scope draw" time.

Also good to see the time from initial open and subsequent. I think the latter is the only one worth looking at, as so much is happening concurrently on initial opening that everything gets bogged down (perhaps something to fix there...?).

MStraeten commented:
I changed static void _lib_histogram_process_waveform according tomyour code snippet

MStraeten commented:
bad news: the pr introduced the risk for segfaults. 
```
* thread #18, name = 'worker res 1', stop reason = EXC_BAD_ACCESS (code=1, address=0x146505ffc)
    frame #0: 0x00000001266f7f55 libhistogram.so`dt_lib_histogram_process at histogram.c:0:41 [opt]
   1   	/*
   2   	    This file is part of darktable,
   3   	    Copyright (C) 2011-2020 darktable developers.
   4
   5   	    darktable is free software: you can redistribute it and/or modify
   6   	    it under the terms of the GNU General Public License as published by
   7   	    the Free Software Foundation, either version 3 of the License, or
Target 0: (darktable) stopped.
(lldb) up
frame #1: 0x00000001266f7c75 libhistogram.so`dt_lib_histogram_process(self=<unavailable>, input=<unavailable>, width=<unavailable>, height=225, in_profile_type=<unavailable>, in_profile_filename=<unavailable>) at histogram.c:301 [opt]
   298 	      _lib_histogram_process_histogram(d, img_display ? img_display : input, width, height);
   299 	      break;
   300 	    case DT_LIB_HISTOGRAM_SCOPE_WAVEFORM:
-> 301 	      _lib_histogram_process_waveform(d, img_display ? img_display : input, width, height);
   302 	      break;
   303 	    case DT_LIB_HISTOGRAM_SCOPE_N:
   304 	      g_assert_not_reached();
(lldb) up
libdarktable.dylib was compiled with optimization - stepping may behave oddly; variables may not be available.
frame #2: 0x000000010029f02a libdarktable.dylib`dt_dev_pixelpipe_process_rec(pipe=<unavailable>, dev=<unavailable>, output=<unavailable>, cl_mem_output=<unavailable>, out_format=<unavailable>, roi_out=<unavailable>, modules=0x0000000119fa6c20, pieces=0x000000012aa54d40, pos=82) at pixelpipe_hb.c:2187:9 [opt]
   2184	      }
   2185	      else
   2186	      {
-> 2187	        darktable.lib->proxy.histogram.process(darktable.lib->proxy.histogram.module, input,
   2188	                                               roi_in.width, roi_in.height,
   2189	                                               darktable.color_profiles->display_type, darktable.color_profiles->display_filename);
   2190	      }
(lldb)
```
sometimes it needs just 5 subsequent "compress history stack" actions, sometimes more then ten - but i encountered it also after several dragging of a slider in a module (which causes a histogram process)

dtorop commented:
> bad news: the pr introduced the risk for segfaults.

That is too bad! I can't reproduce this here. Is there a particular image/XMP which could help? Or might this be another difference between MacOS and Linux builds? If so, it could be a good clue.

I can't tell from the gdb output what the problem is. If you type "bt" does more come up? Or "thread apply all bt" as well, though that'll probably print too much. Or is there a backtrace saved to disk?

I put up another commit which just tries to be safer and more explicit about things. Perhaps worth testing to see if it makes any difference.

For testing timing with/without OpenMP, it would be great to the `final histogram waveform took` and `scope draw took` lines after initial open (I tend to generate these just by scrolling on the histogram to adjust exposure, and take an average of a few, throwing out the high/low). The test would be to use the code in git master (after your Q&D fix), the code in this PR, and then the code in this PR with OpenMP disabled in both process and draw, perhaps something like:

```diff
modified   src/libs/histogram.c
@@ -223,7 +223,7 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
 
   // count the colors
   // FIXME: could flip x/y axes here and when reading to make row-wise iteration?
-#if defined(_OPENMP)
+#if defined(_OPENMP) && !defined(__APPLE__)
 #pragma omp parallel for default(none) \
   dt_omp_firstprivate(in, wf_linear, width, height, wf_width, bin_width, in_stride, out_stride, _height, scale) \
   schedule(static)
@@ -533,7 +533,9 @@ static void _lib_histogram_draw_waveform_channel(dt_lib_histogram_t *d, cairo_t
   for(size_t p = 0; p < nfloats; p += 4)
   {
     const float src = MIN(1.0f, wf_linear[p + ch]);
+#define DT_NO_SIMD_HINTS
     for_four_channels(k,aligned(wf_display,primaries_linear:64))
+#undef DT_NO_SIMD_HINTS
     {
       wf_display[p+k] = src * primaries_linear[ch][k];
     }
@@ -555,7 +557,7 @@ static void _lib_histogram_draw_waveform_channel(dt_lib_histogram_t *d, cairo_t
   // not enough iterations to be worth threading
   for(size_t y = 0; y < wf_height; y++)
   {
-#ifdef _OPENMP
+#if defined(_OPENMP) && !defined(__APPLE__)
 #pragma simd aligned(wf_display, wf_8bit : 64)
 #endif
     for(size_t k = 0; k < wf_width_floats; k++)
```

MStraeten commented:
here three measurements (`darktable -d perf 2>&1 | grep -e 'initing base buffer' -e 'histogram waveform' -e 'scope draw' >perftest*`
1. current master:
[perftest_master.txt](https://github.com/darktable-org/darktable/files/5793165/perftest_master.txt)
2. current master + pr 7767:
[perftest_pr.txt](https://github.com/darktable-org/darktable/files/5793166/perftest_pr.txt)
3. current master + pr 7767 - openmp (according to yout snippet)
[perftest_pr_noomp.txt](https://github.com/darktable-org/darktable/files/5793167/perftest_pr_noomp.txt)

clean build with  `-DCMAKE_BUILD_TYPE=RelWithDebInfo`
Scenario: ~/src/darktable/src/tests/integration/images/mire.1.cr2 with scene referred workflow defaults
4 times: compress history stack
scrolling up and down in histogram

segfaults no longer reproducible ...

dtorop commented:
> segfaults no longer reproducible ...

Great news! It may be that MacOS clang is not letting certain undefined behaviors slide. If so, this is another reason why it's good to write for multiple platforms/compilers.

> here three measurements

Here's a summary (the CPU times aren't meaningful probably due to lots of overlapping processes -- I doctored them heavily by throwing out outlying #'s and taking the median):

step | variant | OpenMP? | time (wall) | time (cpu)
---- | ------- | ------ | ----------- | ----------
process | master | no | 0.004 | 0.009
process | this PR | yes | 0.001 | 0.011
process | this PR | no | 0.006 | 0.014

It's 4x faster to process compared to master, probably mainly because master isn't using OpenMP for MacOS. (And...your computer seems to run this nearly 10x faster than mine!)

step | variant | OpenMP? | time (wall) | time (cpu)
---- | ------- | ------ | ----------- | ----------
draw | master | no | 0.013 | 0.027
draw | this PR | yes | 0.013 | 0.023
draw | this PR | no | 0.013 | 0.021

Draw speed is unchanged on all variants (probably it's limited by Cairo).

What I get from that is that this PR is OK as is for MacOS, and can run with OpenMP enabled. Also, all these times are so short as to make further comparison almost unnecessary.

TurboGit commented on src/libs/histogram.c:
```diff
@@ -208,36 +211,37 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
   // 1.0 is at 8/9 of the height!
   const float _height = (float)(wf_height - 1);
 
+  const size_t in_stride = 4U * width;
```
probably `(size_t)4U * width` needed here. Same below.

ralfbrown replied:
Only if the image width is greater than half a billion pixels :-).

dtorop replied:
Yes, I've been just mulling this! And some of the `size_t` I already wrote is a bit silly, as the output buffer is generally about 360x175 pixels! Code such as b43f62be6c56ed80fc746e109f1672f92ee8e131 looks like a good example, as its legible and should do the right thing.

I understand that so long as the array offsets are `size_t`, it may be good for everything which builds them to be size_t. But then so many more questions. Do I write `for(size_t in_y = 0U; in_y < height; in_y++)` with `0U` because I don't want to risk that a compiler would run-time convert `0` to `0U`? But how do we know `U` is the size of of a `size_t`?

And what if I write `const size_t out_y = isnan(v) ? 0U : CLAMPS((size_t) (v * _height), 0U, _height);` and `v` is negative? What happens when casting a negative number to `size_t` before clamping it... Going down the wormhole of unsigned types.

dtorop replied:
For the record, adding these test lines to the function:

```c
  const int a = 175;
  printf("%zu %zu\n", (size_t)(-1.0f * 175), (size_t)(-1.0f * a));
  printf("%d %zu\n", d->waveform_height, (size_t)(-1.0f * wf_height));
  printf("%zu %zu\n", CLAMPS((size_t)(-1.0f * wf_height), 0U, 175U), (size_t)CLAMPS((-1.0f * wf_height), 0.0f, 175.0f));
```

produces


```
0 0
175 18446744073709551441
175 0
```

I'll think out all these types in a reasonable way and make a PR.

johnny-bit replied:
LOL

generally `(smaller signed type) * (bigger unsigned type)` makes the multiplication work in `smaller signed type` causing potential problems. In this case it's ` (unsigned int) * (signed int)` making whole multiplication work in larger type still...

And floats/doubles are evil too:  `float * int` makes multiplication in float (afaik) but `int *float` and whole multiplication is in int... bah ;) complex stuff. Generally one needs to make sure that multiplication is always done in biggest type unless overflows are intended (and they very hardly ever are) ;)

dtorop replied:
I appreciate the explanation!

Questions, so long as we're doing some explaining, and I don't want to just blindly copy code:
1. Should a loop variable always be `size_t` if it's being used to index an array (or offset a pointer)?
2. Why not `ptrdiff_t`, offsetting pointers is like a pointer difference?
3. What if we know the loop variable will be <= INT_MAX? Should it still be `size_t`? Or `int`, as that's the friendliest type for the processor?
4. What if the loop variable is `int` and it is multiplied by a constant to create an offset. Should it definitely be `size_t` then?
5. If it is multiplied by 4, should that always be written as `4U`?
6. If so, is this for compiler optimization or to make the multiplication work as unsigned?
7. What if it is multiplied by or added to other variables to produce the pointer offset. Should the other variables always be `size_t`?
8. If so, is that to make sure that all the math works in the larger type, or for the sake of optimization, so that they aren't converted to the larger type on the fly?
9. If we know that all the math for the pointer offset will produce a number <= INT_MAX, should all the variables to produce the offset be `int` as that's the friendliest type for the processor?
10. What if the constant is `0`, should it still be written `0U`?

Am going too deep into this, I know...

johnny-bit replied:
Remember: I'm a certified dumbass, whatever I say might be false and/or stupid. Everything i say comes straight from my tiny brain... that being said:

1 & 2. "depends". Usually if loop index is guaranteed to never be below zero and is not used for pointer arithmetics, then `size_t` is perfectly fine. If loop index will be below zero and used for pointer arithmetics then `ptrdiff_t` is recommended.
3. "depends" too - on 64-bit system `int` might be 32 bit while `size_t` will be 64 bit and usually full bit is better.
4 & 5. if index is `int` then simply `int * 4` is fine, if it's `size_t`, then `size_t * 4U` would probably make it more obvious
6. it's more to prevent unwanted compiler behaviour. Usually you don't want multiplication done in smaller type if you know sizes might overflow it. So generally it's better to do multiplication in bigger type
7. "depends". things producing negatives should be in signed type, but if resulting variable shouldn't be negative then variable should be unsigned... The singed/unsigned & type size business bites :confused: 
8. Generally to make sure math is done in largest possible type without overflows.
9. still - depends, int might not be friendlest for processor
10. hmm... `0` is `int`, `0U` is `unsigned`, `0.f` is float, `0.` is double... but 0*whatever = 0 so it doesn't matter :P (except float/double where that's some magic)

I think clang12 will have warns about multiplication type undersizing.

parafin replied:
@johnny-bit, both of your statements are false: https://en.cppreference.com/w/c/language/conversion

Specifically:

- `(smaller signed type) * (bigger unsigned type)` is done in bigger unsigned (or at least int) type
- both `int * float` and `float * int` are done in float

For usual arithmetic operations order of 2 operands doesn't matter for resulting type, but for `a * b * c` you have to remember that first `a * b` will be computed and type of `c` doesn't affect that calculation. So type needs to be manually upgraded if result might not fit.

parafin replied:
Regarding how to write numeric constants (4, 4u, 4.0, 4.0f) - I think it doesn't matter almost any time due to how implicit conversions work, with the exception of using 4.0 in float calculations (which will force it to be done in double). Unless you of course do something intentionally stupid/evil like writing 4.0 in integer calculation.

dtorop replied:
Thanks for all this insight. I pushed another version which makes (I hope) as few changes as possible, and ideally doesn't break MacOS again.

TurboGit reviewed (commented):
A minor comment. This will please @johnny-bit :)

dtorop commented:
And a force push to get rid of an especially absurd "0U".

TurboGit reviewed (approved):
Good to me, thanks!

MStraeten commented:
again bad news:
dt crashes in histogram. i need to bisect to be sure that no later stuff causes it but it looks similar to earlier crashes.
here the backtrace:
```
(darktable:90862): Gtk-CRITICAL **: 18:41:38.949: gtk_window_add_accel_group: assertion 'GTK_IS_WINDOW (window)' failed
libhistogram.so was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 90862 stopped
* thread #18, name = 'worker res 1', stop reason = EXC_BAD_ACCESS (code=1, address=0x13da71ffc)
    frame #0: 0x0000000126dc27eb libhistogram.so`.omp_outlined. at histogram.c:0:58 [opt]
   1   	/*
   2   	    This file is part of darktable,
   3   	    Copyright (C) 2011-2020 darktable developers.
   4
   5   	    darktable is free software: you can redistribute it and/or modify
   6   	    it under the terms of the GNU General Public License as published by
   7   	    the Free Software Foundation, either version 3 of the License, or
Target 0: (darktable) stopped.
(lldb) bt
* thread #18, name = 'worker res 1', stop reason = EXC_BAD_ACCESS (code=1, address=0x13da71ffc)
  * frame #0: 0x0000000126dc27eb libhistogram.so`.omp_outlined. at histogram.c:0:58 [opt]
    frame #1: 0x0000000126dc24fc libhistogram.so`.omp_outlined.(.global_tid.=<unavailable>, .bound_tid.=<unavailable>, wf_width=<unavailable>, bin_width=2, width=<unavailable>, height=<unavailable>, in=0x000000013da72000, in_stride=2680, height_f=<unavailable>, height_i=174, wf_linear=0x0000000126dd0000, out_stride=1340, scale=<unavailable>) at histogram.c:232 [opt]
    frame #2: 0x00000001005fb3d3 libomp.dylib`__kmp_invoke_microtask + 147
    frame #3: 0x00000001005ca9b3 libomp.dylib`__kmp_invoke_task_func + 162
    frame #4: 0x00000001005c7865 libomp.dylib`__kmp_fork_call + 5099
    frame #5: 0x00000001005bed92 libomp.dylib`__kmpc_fork_call + 183
    frame #6: 0x0000000126dbf87e libhistogram.so`dt_lib_histogram_process [inlined] _lib_histogram_process_waveform(d=<unavailable>, input=0x000000013da72000, width=670, height=446) at histogram.c:228:1 [opt]
    frame #7: 0x0000000126dbf75c libhistogram.so`dt_lib_histogram_process(self=<unavailable>, input=<unavailable>, width=<unavailable>, height=<unavailable>, in_profile_type=<unavailable>, in_profile_filename=<unavailable>) at histogram.c:304 [opt]
    frame #8: 0x00000001002d1e4a libdarktable.dylib`dt_dev_pixelpipe_process_rec(pipe=<unavailable>, dev=<unavailable>, output=<unavailable>, cl_mem_output=<unavailable>, out_format=<unavailable>, roi_out=<unavailable>, modules=0x0000000139506ec0, pieces=0x0000000139511e60, pos=81) at pixelpipe_hb.c:2187:9 [opt]
    frame #9: 0x00000001002d0d1e libdarktable.dylib`dt_dev_pixelpipe_process [inlined] dt_dev_pixelpipe_process_rec_and_backcopy(pipe=0x0000000110908200, dev=0x0000000110907000, output=0x000070001152bd80, cl_mem_output=0x000070001152bd88, out_format=0x000070001152bd68, roi_out=0x000070001152bd98, modules=0x0000000139506ec0, pieces=0x0000000139511e60, pos=81) at pixelpipe_hb.c:2254:13 [opt]
    frame #10: 0x00000001002d0ce5 libdarktable.dylib`dt_dev_pixelpipe_process(pipe=0x0000000110908200, dev=<unavailable>, x=<unavailable>, y=<unavailable>, width=670, height=446, scale=5.70004486E-29) at pixelpipe_hb.c:2342 [opt]
    frame #11: 0x0000000100279ac1 libdarktable.dylib`dt_dev_process_preview_job(dev=0x0000000110907000) at develop.c:324:6 [opt]
    frame #12: 0x000000010024c231 libdarktable.dylib`dt_dev_process_preview_job_run(job=<unavailable>) at develop_jobs.c:25:3 [opt]
    frame #13: 0x0000000100246b37 libdarktable.dylib`dt_control_work_res [inlined] dt_control_run_job_res(control=<unavailable>, res=<unavailable>) at jobs.c:215:19 [opt]
    frame #14: 0x00000001002469f4 libdarktable.dylib`dt_control_work_res(ptr=<unavailable>) at jobs.c:519 [opt]
    frame #15: 0x00007fff20355950 libsystem_pthread.dylib`_pthread_start + 224
    frame #16: 0x00007fff2035147b libsystem_pthread.dylib`thread_start + 15
```

MStraeten commented:
the crash occurs after a build using `-DCMAKE_BUILD_TYPE=Release`
if rebuilt with`-DCMAKE_BUILD_TYPE=RelWithDebInfo` it's fine

johnny-bit commented:
How on earth can backtrace contain comment from source file??

MStraeten commented:
don't forget: it's Apple ;)

dtorop commented:
Alas... Just to check, the difference between "Release" and "RelWithDebInfo" is:

Release: -O3 -fexpensive-optimizations
RelWithDebInfo: -O2

I'm still going to blame the code, and not the compiler, but it's confusing.

@MStraeten: I think that you last tested 76e938d55b21c784efc7c5b124f2d6eef284ae01. Do you remember if you tested it before as "Release"? After that, I did one more commit to that code, ddf1975e611467c031cdaae69d07027a037af16f. Perhaps that unsettled things. Would it be possible to compile/test after both these commits and see if there is a difference?

TurboGit commented:
Would be nice to know if it is `-03` or `-fexpensive-optimizations` being the culprit. I would cast my vote for the later.

dtorop commented:
> Would be nice to know if it is -03 or -fexpensive-optimizations being the culprit.

Yes, good point. It would be a last recourse, but it looks like there is a way to turn adjust optimizations for a function. Something like:

```diff
modified   src/libs/histogram.c
@@ -187,6 +187,9 @@ static void _lib_histogram_process_histogram(dt_lib_histogram_t *d, const float
   }
 }
 
+#ifdef __APPLE__
+  __attribute__ ((optimize("2")))
+#endif
 static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *const input, int width, int height)
 {
   dt_times_t start_time = { 0 };
```

MStraeten commented:
i tested with RelWithDebInfo 3 days ago ...

> @MStraeten: I think that you last tested [76e938d](https://github.com/darktable-org/darktable/commit/76e938d55b21c784efc7c5b124f2d6eef284ae01). Do you remember if you tested it before as "Release"? After that, I did one more commit to that code, [ddf1975](https://github.com/darktable-org/darktable/commit/ddf1975e611467c031cdaae69d07027a037af16f). Perhaps that unsettled things. Would it be possible to compile/test after both these commits and see if there is a difference?

i tested both (did a `git checkout 76e938d55b21c784efc7c5b124f2d6eef284ae01` and `ddf1975e611467c031cdaae69d07027a037af16f`) with -`DCMAKE_BUILD_TYPE=Release` - both were fine without a crash
so i'm going to bisect ...

dtorop commented:
> i'm going to bisect

Great. I'll be quite curious for the results. When there is a crash, does it happen every time you run darktable? Or every time you enter darkroom view?

aurelienpierre commented:
`-fexpensive-optimizations` is GCC only IIRC, so it shouldn't even affect Clang.

ralfbrown commented:
> How on earth can backtrace contain comment from source file??

Note that the indicated location is line ZERO of the source file.  So the debugger dutifully showed some lines around that position.

MStraeten commented:
bisecting gave a strange result:  
d71607f303cd0f51acdf402472daea757b80a2c0 is the first bad commit 
but thats: darktableconfig: remove import from now in import module prefs.
nothing that deals with stuff processed in histogram ...
So not a really good scenario.
I spent some time to find a minimum reproducible scenario i can reconstruct before each step of bisecting - just to exclude varying data ... and found:

the best way to get a reproducible crash is 
1. open the image  (mire1.cr2 + default scene referred wokflow + modern color calibration) in darkroom
2. switch between standard histogram and waveform histogram a couple of times 
3. resize the darktable window (enlarging seems to be more efficient to get the crash)
4. switch between standard histogram and waveform histogram a couple of times --> the crash occurs sometimes after the first switch but also after a couple of switches

With this scenario i can reproduce the crash with both commits (`git checkout 76e938d55b21c784efc7c5b124f2d6eef284ae01` and `ddf1975e611467c031cdaae69d07027a037af16f`)
also with `-DCMAKE_BUILD_TYPE=RelWithDebInfo`

bisecting with this scenario gave: `a574a9bcb1bee8c70fd570b0d30f2bd076cf8ded is the first bad commit`

dtorop commented:
It would make sense -- as much as anything does -- if a574a9bcb1bee8c70fd570b0d30f2bd076cf8ded was a problem commit, as it re-enables OpenMP for the processing.

All the steps to reproduce make me wonder if there's a flaw somewhere else, as it's odd that so many UI interactions are needed to trigger the bug. Memory corruption, etc...?

I'm definitely confused about d71607f303cd0f51acdf402472daea757b80a2c0 being pointed to by bisect. That was before you used the reproducible crash scenario?

I'm trying to think out if there were a way to cut out more code interactions to clarify this... Or enforce any assumptions in the waveform process code. If anyone reading this has particular insights/methods, I'm quite interested.

I've been working on the histogram UI (make the buttons be GTK widgets), in the name of getting rid of a lot of hand-coded UI and generally making it easier to add features. I'm wondering if that sort of cleanup would help, but that may be farfetched.

MStraeten commented:
for a further test i replaced both `#ifdef _OPENMP` by `#if defined(_OPENMP) && !defined(__APPLE__) ` but the crash after enlarging the darktable window remains. So at least it seems not to be OpenMP.

MStraeten commented:
i did some test with debug messages in `static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *const input, int width, int height)
`
the crash occurs in the inner loop
```
        for_each_channel(k,aligned(in,wf_linear:16))
        {
          const float v = 1.0f - (8.0f / 9.0f) * in[in_stride * in_y + 4U * in_x + (2U - k)];
          const size_t out_y = isnan(v) ? 0 : MIN((size_t)fmaxf(v*height_f, 0.0f), height_i);
          wf_linear[out_stride * out_y + 4U * out_x + k] += scale;
        }

``` 
in line  `const float v = 1.0f - (8.0f / 9.0f) * in[in_stride * in_y + 4U * in_x + (2U - k)];`
for the 4th channel

dtorop commented:
Interesting. What happens if you skip the 4th channel. This should be one way to do that:

```diff
modified   src/libs/histogram.c
@@ -237,7 +237,9 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
     {
       for(size_t in_y = 0; in_y < height; in_y++)
       {
+#define DT_NO_VECTORIZATION
         for_each_channel(k,aligned(in,wf_linear:16))
+#undef DT_NO_VECTORIZATION
         {
           const float v = 1.0f - (8.0f / 9.0f) * in[in_stride * in_y + 4U * in_x + (2U - k)];
           const size_t out_y = isnan(v) ? 0 : MIN((size_t)fmaxf(v*height_f, 0.0f), height_i);
```

How far through the loop does the crash occur? I'm curious what sort of debug output you were able to find. For my own curiosity, what is output if you do this:

```diff
modified   src/libs/histogram.c
@@ -231,6 +231,7 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
 #endif
   for(size_t out_x = 0; out_x < wf_width; out_x++)
   {
+    printf("bin %ld (%ld total), thread %d/%ld\n", out_x, wf_width, dt_get_thread_num(), dt_get_num_threads());
     const size_t x_from = out_x * bin_width;
     const size_t x_high = MIN(x_from + bin_width, width);
     for(size_t in_x = x_from; in_x < x_high; in_x++)
```

Other thoughts:

Is a `restrict` keyword a problem? Seems unlikely, but to exhaust possibilities, this would remove it:

```diff
modified   src/libs/histogram.c
@@ -193,8 +193,8 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
   if(darktable.unmuted & DT_DEBUG_PERF) dt_get_times(&start_time);
 
   const int wf_height = d->waveform_height;
-  const float *const restrict in = DT_IS_ALIGNED((const float *const restrict)input);
-  float *const restrict wf_linear = DT_IS_ALIGNED((float *const restrict)d->waveform_linear);
+  const float *const in = DT_IS_ALIGNED((const float *const)input);
+  float *const wf_linear = DT_IS_ALIGNED((float *const)d->waveform_linear);
 
   // Use integral sized bins for columns, as otherwise they will be
   // unequal and have banding. Rely on draw to smoothly do horizontal
```

Another question: Is there something weird happening with the input buffer? What if it gets copied into a temporary buffer before processing:

```diff
modified   src/libs/histogram.c
@@ -193,7 +193,8 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
   if(darktable.unmuted & DT_DEBUG_PERF) dt_get_times(&start_time);
 
   const int wf_height = d->waveform_height;
-  const float *const restrict in = DT_IS_ALIGNED((const float *const restrict)input);
+  float *const in = dt_iop_image_alloc(width, height, 4);
+  dt_iop_image_copy_by_size(in, input, width, height, 4);
   float *const restrict wf_linear = DT_IS_ALIGNED((float *const restrict)d->waveform_linear);
 
   // Use integral sized bins for columns, as otherwise they will be
```

dtorop commented:
Oh, I see! It's a silly problem, the `2U - k` on the fourth channel should be the problem. It probably crashes on the 4th channel of the first pixel.

MStraeten commented:
parameters just before crash:
```
fprintf(stderr, "\n %lu: %lu, %lu, %lu" ,k, in_stride,in_y,in_x);
const float v = 1.0f - (8.0f / 9.0f) * in[in_stride * in_y + 4U * in_x + (2U - k)];
```
before crash:
```
 0: 2680, 445, 669
 1: 2680, 445, 669
 2: 2680, 445, 669
 3: 2680, 445, 669

 0: 2680, 0, 0
 1: 2680, 0, 0
 2: 2680, 0, 0
 3: 2680, 0, 0
```

MStraeten commented:
usually the last set doesn't occur just after having resized the darktable window

MStraeten commented:
id did this test with `#if defined(_OPENMP) && !defined(__APPLE__)`

dtorop commented:
> before crash:
> 
> ```
>  0: 2680, 0, 0
>  1: 2680, 0, 0
>  2: 2680, 0, 0
>  3: 2680, 0, 0
> ```

This makes sense, as the code is buggy, at pixel (0,0) it tries to read the pixel before the start of the buffer. I'm not clear why all the UI changes are needed to trigger the bug. Nor why you sometimes don't see that last clause.

What happens if you apply this patch? This *should* make the crash go away.

```diff
modified   src/libs/histogram.c
@@ -237,7 +237,7 @@ static void _lib_histogram_process_waveform(dt_lib_histogram_t *d, const float *
     {
       for(size_t in_y = 0; in_y < height; in_y++)
       {
-        for_each_channel(k,aligned(in,wf_linear:16))
+        for(int k = 0; k < 3; k++)
         {
           const float v = 1.0f - (8.0f / 9.0f) * in[in_stride * in_y + 4U * in_x + (2U - k)];
           const size_t out_y = isnan(v) ? 0 : MIN((size_t)fmaxf(v*height_f, 0.0f), height_i);
```

> id did this test with #if defined(_OPENMP) && !defined(__APPLE__)

That makes sense as even with OpenMP disabled, `for_each_channel()` still looks at all four channels.

https://github.com/darktable-org/darktable/blob/225d9fdaf25855092f32722ce689b19168539226/src/common/darktable.h#L470-L504

dtorop commented:
@MStraeten: does https://github.com/darktable-org/darktable/compare/master...dtorop:waveform-out-of-bounds-fix fix this?

MStraeten commented:
no longer reproducible ... seems to fix it

dtorop commented:
@MStraeten That is great news. I appreciate all the time you put into narrowing this down. Let's hope this is it...

MStraeten commented:
I am better at finding errors than I am at fixing them ... ;)

dtorop commented:
And it's hard to fix bugs until they're found...
